pkg/serverless/trace/cold_start_span_creator.go

```diff
@@ -35,45 +33,90 @@ const (
 var functionName = os.Getenv(functionNameEnvVar)
 
 type ColdStartSpanCreator struct {
-	executionContext *executioncontext.ExecutionContext
-	traceAgent       *ServerlessTraceAgent
-	createSpan       *sync.Once
+	TraceAgent            *ServerlessTraceAgent
+	CreateSpan            *sync.Once
+	LambdaSpanChan        chan *pb.Span
+	InitDurationChan      chan float64
+	SyncSpanDurationMutex *sync.Mutex
+	ColdStartSpanId       uint64
+	lambdaSpan            *pb.Span
+	initDuration          float64
+	StopChan              chan struct{}
 }
 
-func (c *ColdStartSpanCreator) create(lambdaSpan *pb.Span) {
-	// Prevent infinite loop from SpanModifier call
-	if lambdaSpan.Name == "aws.lambda.cold_start" {
+func (c *ColdStartSpanCreator) Run() {
+	go func() {
+		for {
+			select {
+			case span := <-c.LambdaSpanChan:
+				if span.Name == spanName {
+					continue
+				}
+				c.lambdaSpan = span
```

Comment:
I do not think this is thread safe. We'll end up writing and reading the `lambdaSpan` field from different goroutines.

Reply:
Hmmm I think it's the same goroutine, and since both methods call createIfReady, which locks, that should be safe.

But I think I can refactor to be more clear